### ISSUE TITLE
Installing from fixed RSPM snapshots only

### ISSUE THREAD

**Issue by pat-s** [rstudio/renv#531]
When trying to install from a fixed RSPM snapshot only, {renv} updates the "Repository" field for some packages in `renv.lock` from RSPM to some CRAN refs, even though no CRAN repo is listed in `options(repo = )`.

I assume this happens due to cache linking? I.e. if the installed matched version requested was installed from CRAN on the local machine, it will be used instead of downloading the RSPM binary?

However, if a different user now calls `renv::restore()`, restoring will fail because there is no CRAN repo configured which `renv.lock` refers to.

It would be nice if only entries which are listed in `options(repo = )` were allowed to be valid entries for the "Repository" metadata field in `renv.lock`.

The following `.Rprofile` can be used to mimic this situation in a project:

```r
source("renv/activate.R")
options(renv.config.auto.snapshot = TRUE)
options(repos = structure(c(
  RSPM = "https://packagemanager.rstudio.com/cran/301"
)))
```

**Comment by pat-s** [comment]
It looks like that in practice (i.e. when restoring) the metadata field in `renv.lock` has no relevance. `renv` will only use the repos which are configured in `.Rprofile` for restoring.

Even though this might be a potential point for confusion it does not affect usability :)

**Comment by kevinushey** [comment]
That's right. One challenge here is that packages from RSPM still "declare" themselves as packages from CRAN; that is, they still have

```
Repository: CRAN
```

in their `DESCRIPTION` file. We could adjust this so that `renv` updates the `DESCRIPTION` at install time, but the downside with this is that it won't catch cases where a package was installed from RSPM but not by `renv`.

**Comment by slopp** [comment]
To me, it feels like this is a core question about what belongs in a lock file vs user preferences. 

Right now, the repo option is a user preference. I believe if you shared your renv options in addition to the lockfile, the restore would do what you want without sharing a. .Rprofile.

However, I'm curious about the motivation? Since renv.lock specifies the exact versions you want, it seems like overkill to also specify the repo snapshot. Is there another motivator for wanting the project to "restore" this repo option?

**Comment by pat-s** [comment]
> I believe if you shared your renv options in addition to the lockfile, the restore would do what you want without sharing a. .Rprofile.

I don't understand this part.

> However, I'm curious about the motivation? 

I want to set a fixed RSPM snapshot for the lifetime of a project with respect to CRAN package versions.
Important updates should refer to GH snapshots.
Otherwise, the project should run on a fixed RSPM snapshot and an `install.packages()` should always only install this one version of the snapshot time.

**Comment by krlmlr** [comment]
One issue we're facing is that the `Repository` field in `renv.lock` alternates between `RSPM` and `CRAN` in a seemingly random manner. This occurs in projects with multiple collaborators. We suspect that this might be linked to how the package was installed to the renv cache, we don't have a reprex yet.

Do we even need the `Repository` field in `renv.lock`?

**Comment by kevinushey** [comment]
The next time you see this occur, could you share the contents of the DESCRIPTION file for the snapshotted package? That is:

```
system(paste("cat", system.file("DESCRIPTION", package = <pkg>)))
```

**Comment by kevinushey** [comment]
It might also depend on how the repositories are named for a session; e.g. if you've set something like:

```
options(repos = c(RSPM = "https://packagemanager.rstudio.com/all/latest"))
```

then I think `renv` _may_ use the repository name to infer the Repository field for the lockfile.

That said, while the repository field is currently unused, in theory we might want to use it in case the user had:

1. Multiple repositories configured,
2. More than one of those repositories offered a particular version of a package,
3. The user wanted to ensure we retrieved the package from repo A rather than repo B.

A niche case, to be fair.

It's also useful (when it works) for documentation at which repository provides a particular package, when multiple repositories are being used.

**Comment by krlmlr** [comment]
This is still an issue with our projects, the Git history is cluttered with these changes.

Would you mind substituting RSPM -> CRAN when updating the lockfile?

**Comment by kevinushey** [comment]
I believe this is caused by this part of the code:

https://github.com/rstudio/renv/blob/5507f5f9726df109098562104305232f1de911ed/R/snapshot.R#L664-L665

That is, we preserve the `Repository:` field from the `DESCRIPTION` file (if present) when updating the lockfile.

Would you be open to a somewhat more kludgey solution, e.g. a user-defined hook (as an R option) that either remaps the repository field here as appropriate, or perhaps more globally just takes the lockfile (as an R object) and returns a suitably-mutated version on write?

**Comment by krlmlr** [comment]
## TL;DR

A hook is probably fine, the first thing I'd do is to scrub the `"Repository"` field altogether.

## Perhaps change the behavior?

I think I see the problem now.

- I'm running Linux, obtaining packages from RSPM. This sends over `DESCRIPTION` files containing `Repository: RSPM` .
- My teammate runs macOS, obtaining packages from MRAN. That repo sends `DESCRIPTION` files containing `Repository: CRAN` .

Not verified yet, but this feels consistent with our use case and with the expected behavior from that.

I see a problem with using a field from the `DESCRIPTION` that is populated during *installation* time, where we actually want to record a *feasible package source* in `renv.lock`. There is no guarantee that this will be stable.

I'd argue that we don't even need to record the repository if `renv.lock` contains only one cranlike. In the case of multiple cranlikes, we should really record the repository from where a package can be obtained if we really need it. Where does renv even use that information?

Could we perhaps record a list of available packages in a separate file under `renv/` ?

**Comment by hadley** [comment]
Duplicate of #1001 / #1175
